include/nix/DataFrame.hpp

```diff
@@ -0,0 +1,142 @@
+public:
+
+    DataFrame()
+        : EntityWithSources()
+        {}
+
+    DataFrame(const std::shared_ptr<base::IDataFrame> &df)
+        : EntityWithSources(df)
+        {}
+
+    DataFrame(std::shared_ptr<base::IDataFrame> &&ptr)
+        : EntityWithSources(std::move(ptr))
+        {}
+
+    ndsize_t rows() const {
+        return backend()->rows();
+    }
+
+    void rows(ndsize_t n) {
+        return backend()->rows(n);
+    }
+
+    virtual std::vector<unsigned> nameToCol(const std::vector<std::string> &names) const {
+        return backend()->nameToCol(names);
+    }
+
+    virtual std::vector<std::string> colToName(const std::vector<unsigned> &cols) const {
+        return backend()->colToName(cols);
+    }
+
+    std::vector<Column> columns() const {
+        return backend()->columns();
+    }
+
+    void writeRow(ndsize_t row, const std::vector<Variant> &v) {
+        return backend()->writeRow(row, v);
+    }
+
+    void writeCell(ndsize_t row, unsigned col, const Variant &v) {
+        return this->writeCells(row, {{col, v}});
+    }
+
+    void writeCells(ndsize_t row, const std::vector<Cell> &cells) {
+        return backend()->writeCells(row, cells);
+    }
+
+    Variant readCell(ndsize_t row, unsigned col) const {
+        const std::vector<std::string> &cols = colToName({col});
+        if (cols.size() < 1) {
+            return Variant{};
+        }
+        return this->readCell(row, cols[0]);
+    }
+
+    Variant readCell(ndsize_t row, const std::string &col) const {
+        std::vector<Variant> cells = backend()->readCells(row, {col});
+        if (cells.size() < 1) {
+            return Variant{};
+        }
+
+        return cells[0];
+    }
+
+    std::vector<Variant> readCells(ndsize_t row, const std::vector<std::string> &cols) const {
+        return backend()->readCells(row, cols);
+    }
+
+    std::vector<Variant> readRow(ndsize_t row) {
+        return backend()->readRow(row);
+    }
+
+    template<typename T>
+        void writeColumn(int col,
+                         ndsize_t offset,
+                         const std::vector<T> &vals,
+                         ndsize_t count = 0) {
+        const Hydra<const std::vector<T>> hydra(vals);
+
+        if (count == 0)
+            count = vals.size();
+        else if (count > vals.size())
+            throw OutOfBounds("Requested to write more data than available");
+
+        DataType dtype = hydra.element_data_type();
+        backend()->writeColumn(col, offset, count, dtype, (const char *) hydra.data());
+    }
+
+    template<typename T>
+        void readColumn(int col,
+                        ndsize_t offset,
+                        std::vector<T> &vals,
+                        ndsize_t count = 0) {
+        Hydra<std::vector<T>> hydra(vals);
+
+        if (count == 0)
+            count = vals.size();
+
+        DataType dtype = hydra.element_data_type();
+        backend()->readColumn(col, offset, count, dtype, (void *) hydra.data());
+    }
+
```

Comment:
API: I guess we could expose the raw `readColumn`, `writeColumn` functions too? Or we can create a custom data structure that takes a raw pointer, length and datatype as a proxy structure. Might be nix also for DataArrays. Or both.